api/src/campaigns/campaigns.service.ts

```diff
@@ -0,0 +1,372 @@
+//       {
+//         campaign: campaign.id,
+//       },
+//       {
+//         campaign: campaign.id,
+//       },
+//     )
+
+//     const data = p2v.data || {}
+//     const viability = data.viability || {}
+//     const updatedData = {
+//       ...data,
+//       viability: {
+//         ...viability,
+//         [columnKey]: value,
+//       },
+//     }
+
+//     await PathToVictory.updateOne({ id: p2v.id }).set({
+//       data: updatedData,
+//     })
+
+//     if (!campaign.pathToVictory) {
+//       await Campaign.updateOne({ id: campaign.id }).set({
+//         pathToVictory: p2v.id,
+//       })
+//     }
+//   } catch (e) {
+//     console.log('Error at updateViability', e)
+//     await sails.helpers.slack.errorLoggerHelper('Error at updateViability', e)
+//   }
+// }
+
+function buildQueryWhereClause({
+  id,
+  state,
+  slug,
+  email,
+  level,
+  primaryElectionDateStart,
+  primaryElectionDateEnd,
+  campaignStatus,
+  generalElectionDateStart,
+  generalElectionDateEnd,
+  p2vStatus,
+}) {
+  return `
+  ${id ? ` AND c.id = ${id}` : ''}
+  ${slug ? ` AND c.slug ILIKE '%${slug}%'` : ''}
+  ${email ? ` AND u.email ILIKE '%${email}%'` : ''}
+  ${state ? ` AND c.details->>'state' = '${state}'` : ''}
+  ${level ? ` AND c.details->>'ballotLevel' = '${level.toUpperCase()}'` : ''}
+  ${
+    campaignStatus
+      ? ` AND c.is_active = ${campaignStatus === 'active' ? 'true' : 'false'}`
+      : ''
+  }
+  ${
+    primaryElectionDateStart
+      ? ` AND c.details->>'primaryElectionDate' >= '${primaryElectionDateStart}'`
+      : ''
+  }
+  ${
+    primaryElectionDateEnd
+      ? ` AND c.details->>'primaryElectionDate' <= '${primaryElectionDateEnd}'`
+      : ''
+  }
+  ${
+    generalElectionDateStart
+      ? ` AND c.details->>'electionDate' >= '${generalElectionDateStart}'`
+      : ''
+  }
+  ${
+    generalElectionDateEnd
+      ? ` AND c.details->>'electionDate' <= '${generalElectionDateEnd}'`
+      : ''
+  }
+  ${p2vStatus ? ` AND p.data->>'p2vStatus' = '${p2vStatus}'` : ''}
+`
+}
+
+function buildCustomCampaignListQuery({
+  id,
+  state,
+  slug,
+  email,
+  level,
+  primaryElectionDateStart,
+  primaryElectionDateEnd,
+  campaignStatus,
+  generalElectionDateStart,
+  generalElectionDateEnd,
+  p2vStatus,
+}: Partial<CampaignListQuery>) {
+  console.log(generalElectionDateStart)
+
+  return `
+  SELECT
+    c.*,
+    u.first_name as "first_name",
+    u.last_name as "last_name",
+    u.phone as "phone",
+    u.email as "email",
+    u.meta_data,
+    p.data as "pathToVictory"
+  FROM public.campaign AS c
+  JOIN public.user AS u ON u.id = c.user_id
+  LEFT JOIN public.path_to_victory as p ON p.campaign_id = c.id
+  WHERE c.user_id IS NOT NULL
+  ${buildQueryWhereClause({
+    id,
+    state,
+    slug,
+    email,
+    level,
+    primaryElectionDateStart,
+    primaryElectionDateEnd,
+    campaignStatus,
+    generalElectionDateStart,
+    generalElectionDateEnd,
+    p2vStatus,
+  })}
+  ORDER BY c.id DESC;
+`
+}
```

Comment:
Doesn't _have_ to be done now.. But, can we add some `TODO`s here to look into seeing if it 's possible to use Prisma's query builder for these now? Waterline didn't play nicely w/ `JSONb` columns, which was why we dropped down toe SQL for these.

But it looks like Prisma can handle JSONb fields just fine in it's query builder: https://www.prisma.io/docs/orm/prisma-client/special-fields-and-types/working-with-json-fields

Reply:
I'm gonna go ahead and approve.. I'm fine if we just merge w/ a `TODO` comment on these raw SQL methods.. but even better if you wanna see if we can indeed build these w/ the query builder. Your call. 👍